libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,312 @@
+			if (_semi[_label[uIdx]] < _semi[_label[_vIdx]])
+				_label[_vIdx] = _label[uIdx];
+			_ancestor[_vIdx] = _ancestor[uIdx];
+		}
+		solAssert(_label[uIdx] <= _label[_vIdx]);
+	}
+
+	std::vector<size_t> findDominators(V const& _entry, size_t numVertices)
+	{
+		solAssert(numVertices > 0);
+		// semi(w): The DFS index of the semidominator of ``w``.
+		std::vector<size_t> semi(numVertices, std::numeric_limits<size_t>::max());
+		// parent(w): The index of the vertex which is the parent of ``w`` in the spanning
+		// tree generated by the DFS.
+		std::vector<size_t> parent(numVertices, std::numeric_limits<size_t>::max());
+		// ancestor(w): The highest ancestor of a vertex ``w`` in the dominator tree used
+		// for path compression.
+		std::vector<size_t> ancestor(numVertices, std::numeric_limits<size_t>::max());
+		// label(w): The index of the vertex ``w`` with the minimum semidominator in the path
+		// to its parent.
+		std::vector<size_t> label(numVertices, 0);
+
+		// ``link`` adds an edge to the virtual forest.
+		// It copies the ``parent`` index of the vertex ``w`` to the ancestor array to limit the search path upwards.
+		// TODO: implement sophisticated link-eval algorithm as shown in pg 132
+		// See: https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+		auto link = [&](size_t _parentIdx, size_t _wIdx)
+		{
+			ancestor[_wIdx] = _parentIdx;
+		};
+
+		// ``eval`` computes the path compression.
+		// Finds ancestor with lowest semi-dominator DFS number (i.e. index).
+		auto eval = [&](size_t _vIdx) -> size_t
+		{
+			if (ancestor[_vIdx] != std::numeric_limits<size_t>::max())
+			{
+				compressPath(ancestor, label, semi, _vIdx);
+				return label[_vIdx];
+			}
+			return _vIdx;
+		};
+
+		auto toIdx = [&](V const& v) { return m_vertexIndices[v]; };
+
+		// step 1
+		std::set<V> visited;
+		// predecessors(w): The set of vertices ``v`` such that (``v``, ``w``) is an edge of the graph.
+		std::vector<std::set<size_t>> predecessors(numVertices);
+		// bucket(w): a set of vertices whose semidominator is ``w``
+		// The index of the array represents the vertex's ``dfIdx``
+		std::vector<std::deque<size_t>> bucket(numVertices);
+		// idom(w): the index of the immediate dominator of ``w``
+		std::vector<size_t> idom(numVertices, std::numeric_limits<size_t>::max());
+		// The number of vertices reached during the DFS.
+		// The vertices are indexed based on this number.
+		size_t dfIdx = 0;
+		auto dfs = [&](V const& _v, auto _dfs) -> void {
+			auto [_, inserted] = visited.insert(_v);
+			if (!inserted)
+				return;
+			m_vertices[dfIdx] = _v;
+			m_vertexIndices[_v] = dfIdx;
+			semi[dfIdx] = dfIdx;
+			label[dfIdx] = dfIdx;
+			dfIdx++;
+			ForEachSuccessor{}(_v, [&](V const& w) {
+				if (semi[dfIdx] == std::numeric_limits<size_t>::max())
+				{
+					parent[dfIdx] = m_vertexIndices[_v];
+					_dfs(w, _dfs);
+				}
+				predecessors[m_vertexIndices[w]].insert(m_vertexIndices[_v]);
+			});
```

Comment:
Something seems wrong here. I think `if (semi[dfIdx] == std::numeric_limits<size_t>::max())` will always be true, because we are going through elements of `semi` in order, without backtracking.

This also means that `parent[dfIdx]` is getting overwritten for each successor. Shouldn't the condition be checking `parent` rather than `semi`?

Reply:
No, `semi[dfIdx]` is set to `dfIdx` for each new visited vertex (i.e. `_v`) (see line [222](https://github.com/ethereum/solidity/pull/14460/files#diff-dd15dbb6cf69e9d36829f239cc9b13ae4af967aa132114cc722eb1b01d0f4f7dR222) above). `dfIdx` is incremented at line `224`, before iterate over the successors, so the condition in `ForEachSuccessor` will be false as soon you visit all vertices at least once. The idea is to assign unique IDs (the `dfIdx`) to all vertices based on their DFS order of visiting (see Appendix B of https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf for the pseudo-code of this algorithm). But note that here we start from `dfIdx` being 0 and not 1, therefore I used `max()` instead of 0 as done in the paper).

The `ForEachSuccessor` iterates over the successors of each visited vertex in the graph and if the successor vertex (i.e. `w`) was not visited (i.e. `semi[dfIdx]` is `std::numeric_limits<size_t>::max()`) it adds the current visited vertex (i.e. `_v`) as parent of `w`, which will be assigned to his `dfIdx` in the recursive call. It is done in this way, because, when visiting the vertices, I don't know yet their `dfIdx`.  We could replace the condition by:
```c++
if (!visited.count(w))
```
if you think it makes things clear. I thought that it would be clear if I used the `semi[dfIdx]` because it would show that was the same index used when assigning the parent and that will be set by the recursive call. But yeah, maybe it was more confusing. So probably a comment mentioning that `dfIdx` in the condition is the index of `w` (which will be assigned to `semi[dfIdx]` in the recursion) should clarify the things?

> And shouldn't we be setting the parent of m_vertexIndices[w]?

We are, but based on the DFS index of `w` when discovering the vertex, which will be `dfIdx` when that `if` statement is true, after which, it should never be true again for the same `w`.

Reply:
Ah, I see it now. What confused me was that I somehow missed that `dfIdx` is not a local variable in `dfs()` and its value will be different for each cycle in the successor loop. Without this it looked like its value would stay the same in each cycle. Now it makes sense.

The `visited.count(w)` condition is still clearer though so I'd keep it. You could just add a comment that in each cycle `dfIdx` is the index of vertex `w`.

Reply:
I added a comment there ;)